aten/src/ATen/native/transformers/sdp_utils_cpp.cpp

```diff
@@ -0,0 +1,163 @@
+#include <ATen/native/transformers/sdp_utils_cpp.h>
+
+namespace sdp {
+namespace {
+
+std::array<SDPBackend, num_backends> priority_order_cpp(sdp_params params) {
+  constexpr std::array<SDPBackend, num_backends> default_order{
+      SDPBackend::flash_attention,
+      // SDPBackend::efficient_attention,
+      SDPBackend::math};
+
+  return default_order;
+}
+
+bool check_head_dim_size_cpp(sdp_params params, bool debug) {
+  const auto query_size_last = params.query.sym_size(-1);
+  const auto key_size_last = params.key.sym_size(-1);
+  const auto value_size_last = params.value.sym_size(-1);
+  if (!(query_size_last == key_size_last &&
+        query_size_last == value_size_last)) {
+    if (debug) {
+      TORCH_WARN(
+          "Flash attention requires q,k,v to have the same last dimension.",
+          " Got Query.size(-1): ",
+          query_size_last,
+          ", Key.size(-1): ",
+          params.key.sym_size(-1),
+          ", Value.size(-1): ",
+          params.value.sym_size(-1),
+          " instead.");
+    }
+    return false;
+  }
+  return true;
+}
+
+bool check_head_dim_size_mem_efficient_cpp(sdp_params params, bool debug) {
+  const auto query_size_last = params.query.sym_size(-1);
+  const auto value_size_last = params.value.sym_size(-1);
+  const int64_t alignment = 1;
+  if (!(query_size_last == params.key.sym_size(-1) &&
+        query_size_last % alignment == 0 && query_size_last > 0 &&
+        value_size_last % alignment == 0 && value_size_last > 0)) {
+    if (debug) {
+      TORCH_WARN(
+          "Mem efficient attention requires last dimension of inputs to be divisible by ",
+          alignment,
+          ". ",
+          "Got Query.size(-1): ",
+          query_size_last,
+          ", Key.size(-1): ",
+          params.key.sym_size(-1),
+          ", Value.size(-1): ",
+          params.value.sym_size(-1),
+          " instead.");
+    }
+    return false;
+  }
+  return true;
+}
+
+bool use_flash_attention_cpp(sdp_params params, bool debug) {
+  constexpr auto cpp_supported_flash_dtypes =
+      array_of<at::ScalarType>(at::kFloat, at::kDouble, at::kBFloat16);
+
+  // Define gate functions that determine if a flash kernel can be run
+  constexpr auto constraints = array_of<bool (*)(sdp_params, bool)>(
+      check_runtime_disabled_flash,
+      check_nested_tensor,
+      check_for_dropout,
+      check_tensor_shapes,
+      check_batch_size_and_num_heads,
+      check_for_attn_mask,
+      check_head_dim_size_cpp,
+      check_nonzero_sequence_lengths,
+      check_last_dim_stride_equals_1);
+  for (auto& constraint : constraints) {
+    if (!constraint(params, debug)) {
+      return false;
+    }
+  }
+
+  return check_tensor_dtype(params, cpp_supported_flash_dtypes, debug);
+}
+
+bool use_mem_efficient_attention_cpp(sdp_params params, bool debug) {
```

Comment:
If we aren't going to implement this we should just remove

Reply:
And all the commented out parts as well

Reply:
Thanks and modified.